src/appmixer/kafka/connections.js

```diff
@@ -0,0 +1,204 @@
+'use strict';
+
+const RegexParser = require('regex-parser');
+const { kafka } = require('./common.js');
+
+let openConnections = {};
+let inProgress = false;
+const callbacks = [];
+
+const initializeKafkaConsumer = ({ groupId, authDetails }) => {
+
+    const kafkaMaster = kafka();
+    kafkaMaster.init(authDetails);
+    return kafkaMaster.createConsumer({ groupId });
+};
+
+const initializeKafkaProducer = (authDetails) => {
+
+    const kafkaMaster = kafka();
+    kafkaMaster.init(authDetails);
+    return kafkaMaster.createProducer();
+};
+
+const processMessageHeaders = (headers) => {
+
+    const processedHeaders = {};
+    if (headers) {
+        Object.entries(headers).forEach(([key, value]) => {
+            processedHeaders[key] = Buffer.isBuffer(value) ? value.toString('utf8') : value || '';
+        });
+    }
+    return processedHeaders;
+};
+
+const processMessageData = (message) => {
+
+    return {
+        key: message.key.toString(),
+        value: message.value.toString(),
+        headers: processMessageHeaders(message.headers)
+    };
+};
+
+const addConnection = async (context, component, mode) => {
+
+    const { topics, flowId, componentId, fromBeginning, authDetails, groupId } = component;
+    authDetails.connectionTimeout = context.config.connectionTimeout;
+
+    const connectionId = `${flowId}:${componentId}`;
+
+    // If another connection operation is in progress, wait for it to complete
+    if (inProgress) {
```

Comment:
has to be `inProgress[connectionId]`. Otherwise, you would combine multiple connections (from multiple SendMessage components. But adding connections for different `connectionId` can happen simultaneously.